src/sampling.jl

```diff
@@ -42,6 +42,47 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
+# check whether we can use T to store indices 1:n exactly
+for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
+    @eval isexact(n, ::Type{$T}) = true
+end
+isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
+isexact(n, ::Type{BigInt}) = false # bigint is too expensive
+for T in (Float32, Float64)
```

Comment:
Why not `Float16` too? Also, shouldn't there be a fallback definition for `AbstractFloat` just like for `Integer`?

BTW, have you checked performance? I wonder whether it really makes sense to store integer indices e.g. in a `Vector{Complex{Float64}}`. The memory savings may not be worth it if it's slower.

Reply:
I was worried that `Float16` might be too slow, and similarly for other `AbstractFloat` types; `Float32` and `Float64` are the only two that generally have hardware support.

Reply:
I ran some benchmarks, and (as you might expect) it depends on the length of `x`.

If you are only computing a few samples (e.g. 5), then it is faster to avoid the allocation.   If you are computing lots of samples (e.g. 1000) it's faster to allocate the indices.

I guess I could hard-code an approximate crossover point.